radixdlt/src/main/java/com/radixdlt/consensus/BFTEventPreprocessor.java

```diff
@@ -0,0 +1,223 @@
+	private final PacemakerState pacemakerState;
+	private final ProposerElection proposerElection;
+	private final SystemCounters counters;
+	private final SyncQueues queues;
+
+	public BFTEventPreprocessor(
+		ECPublicKey myKey,
+		BFTEventProcessor forwardTo,
+		PacemakerState pacemakerState,
+		VertexStore vertexStore,
+		ProposerElection proposerElection,
+		SyncQueues queues,
+		SystemCounters counters
+	) {
+		this.myKey = Objects.requireNonNull(myKey);
+		this.pacemakerState = Objects.requireNonNull(pacemakerState);
+		this.vertexStore = Objects.requireNonNull(vertexStore);
+		this.proposerElection = Objects.requireNonNull(proposerElection);
+		this.queues = queues;
+		this.counters = Objects.requireNonNull(counters);
+		this.forwardTo = forwardTo;
+	}
+
+	private String getShortName() {
+		return myKey.euid().toString().substring(0, 6);
+	}
+
+	private boolean peekAndExecute(SyncQueue queue, Hash vertexId) {
+		final HasSyncConsensusEvent event = queue.peek(vertexId);
+		if (event == null) {
+			return false;
+		}
+
+		// Explicitly using switch case method here rather than functional method
+		// to process these events due to much better performance
+		if (event instanceof NewView) {
+			final NewView newView = (NewView) event;
+			return this.processNewViewInternal(newView, false);
+		}
+
+		if (event instanceof Proposal) {
+			final Proposal proposal = (Proposal) event;
+			return this.processProposalInternal(proposal, false);
+		}
+
+		throw new IllegalStateException("Unexpected consensus event: " + event);
+	}
+
+	/**
+	 * Signal that vertexStore now contains the given vertexId.
+	 * Executes events which are dependent on this vertex
+	 *
+	 * @param vertexId the id of the vertex which is now guaranteed be synced.
+	 */
+	public void processLocalSync(Hash vertexId) {
+		for (SyncQueue queue : queues.getQueues()) {
+			boolean first = true;
+			while (peekAndExecute(queue, first ? vertexId : null)) {
+				queue.pop();
+				first = false;
+			}
+		}
+
+		forwardTo.processLocalSync(vertexId);
+	}
+
+	@Override
+	public void processVote(Vote vote) {
+		log.trace("{}: VOTE: PreProcessing {}", this.getShortName(), vote);
+
+		// only do something if we're actually the leader for the vote
+		final View view = vote.getVoteData().getProposed().getView();
+		// TODO: currently we don't check view of vote relative to our pacemakerState. This opens
+		// TODO: up to dos attacks on calculation of next proposer if ProposerElection is
+		// TODO: an expensive operation. Need to figure out a way of mitigating this problem
+		// TODO: perhaps through filter views too out of bounds
+		if (!Objects.equals(proposerElection.getProposer(view), myKey)) {
+			log.warn("{}: VOTE: Ignoring confused vote {} for {}",
+				getShortName(), vote.hashCode(), vote.getVoteData().getProposed().getView());
+		}
+
+		forwardTo.processVote(vote);
+	}
+
+	private boolean processNewViewInternal(NewView newView, boolean enqueueIfFail) {
```

Comment:
It seems redundant to have the `enqueueIfFail` flag as a parameter here, when all it's used for is to act on the return value. Might be cleaner to move the automatic enqueuing to the call site where it's actually required (down in `processNewView`). That is, change line 165 to something like `if (!processNewViewInternal(newView)) { // enqueue }`

Reply:
The same goes for `processProposalInternal`

Reply:
good point